osu.Game/Screens/OnlinePlay/OnlinePlaySongSelect.cs

```diff
@@ -124,36 +124,43 @@ protected override void LoadComplete()
 
         private void onFreestyleChanged(ValueChangedEvent<bool> enabled)
         {
+            // Remove invalid mods and display the newly available mod panels.
+            Mods.Value = Mods.Value.Where(isValidGlobalMod).ToArray();
+            ModSelect.IsValidMod = isValidGlobalMod;
+            FreeMods.Value = FreeMods.Value.Where(isValidFreeMod).ToArray();
+            freeModSelect.IsValidMod = isValidFreeMod;
+
             if (enabled.NewValue)
             {
                 freeModsFooterButton.Enabled.Value = false;
-                freeModsFooterButton.Enabled.Value = false;
-                ModsFooterButton.Enabled.Value = false;
-
-                ModSelect.Hide();
                 freeModSelect.Hide();
 
-                Mods.Value = [];
+                // Freestyle allows all mods to be selected as freemods. This does not play nicely for some components:
+                // - We probably don't want to store a gigantic list of acronyms to the database.
+                // - The mod select overlay isn't built to handle duplicate mods/mods from all rulesets being shoved into it.
+                // Instead, freestyle inherently assumes this list is empty, and must be empty for server-side validation to pass.
                 FreeMods.Value = [];
             }
             else
             {
                 freeModsFooterButton.Enabled.Value = true;
-                ModsFooterButton.Enabled.Value = true;
+
+                // When disabling freestyle, enable freemods by default.
+                FreeMods.Value = freeModSelect.AllAvailableMods.Where(state => state.ValidForSelection.Value).Select(state => state.Mod).ToArray();
```

Comment:
This sort of breaks an existing interaction:

- Pick a playlist item with freestyle off but any free mods specified
- Edit the playlist item
- Free mods will deselect automatically

Probably something or other about reading `AllAvailableMods` too soon before it's populated or something.

This is the closest thing to a blocker on this PR that I've found but not sure whether to require fixing it here or do it after.

Reply:
I'm cautiously aware that there's no tests for this, but I've attempted to fix in https://github.com/ppy/osu/pull/32674/commits/43386a193b6f5da2eb3a4efa7e2e537b60b42238.

And yeah, part of it is related to `AllAvailableMods` being read too soon, but even fixing that won't actually make things behave appropriately. It really needs to not run the `onFreestyleChanged()` logic on load so that it doesn't interfere with the item's state.